### PR TITLE
chore(ts): Convert app/data/forms

### PR DESCRIPTION
**Pull request by evanpurkhiser**
Makes some additions to the settings form types

### CODE REVIEW

File: src/sentry/static/sentry/app/views/settings/components/forms/type.tsx [hunk 1]
```diff
@@ -42,26 +46,30 @@ type BaseField = {
...
-  formatLabel?: (value: number | '') => React.ReactNode;
+  choices?:
+    | ((props: {[key: string]: any}) => void)
+    | readonly Readonly<[number | string, React.ReactNode]>[];
 
+  formatLabel?: (value: number | '') => React.ReactNode;
+  transformInput?: (value: FieldValue) => FieldValue;
+  getData?: (data: object) => object;
+  /**
+   * ff false, disable saveOnBlur for field, instead show a save/cancel button
```
Comment (EvanPurkhiser) [inline src/sentry/static/sentry/app/views/settings/components/forms/type.tsx]:
```suggestion
   * If false, disable saveOnBlur for field, instead show a save/cancel button
```

File: src/sentry/static/sentry/app/data/forms/accountNotificationSettings.tsx [hunk 2]
```diff
@@ -1,9 +1,10 @@
 import {t, tct} from 'app/locale';
+import {Field, JsonFormObject} from 'app/views/settings/components/forms/type';
 
 // Export route to make these forms searchable by label/help
 export const route = '/settings/account/notifications/';
 
-export const fields = {
+export const fields: {[key: string]: Field} = {
```
Comment (markstory) [inline src/sentry/static/sentry/app/data/forms/accountNotificationSettings.tsx]: We have a `StringMap` generic in app/types that would make this `StringMap<Field>`

File: src/sentry/static/sentry/app/data/forms/projectGeneralSettings.tsx [hunk 3]
```diff
@@ -95,14 +96,17 @@ export const fields = {
       "Automatically resolve an issue if it hasn't been seen for this amount of time"
     ),
     formatLabel: val => {
-      val = parseInt(val, 10);
+      val = Number(val);
```
Comment (markstory) [inline src/sentry/static/sentry/app/data/forms/projectGeneralSettings.tsx]: This will behave a bit different if the input value is float shaped.

**Review by markstory — APPROVED** [review]
Looks good to me. The Number() change could result in different behavior though.

File: src/sentry/static/sentry/app/data/forms/projectGeneralSettings.tsx (hunk 3 above)
Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/data/forms/projectGeneralSettings.tsx]: This should be OK

File: src/sentry/static/sentry/app/data/forms/accountNotificationSettings.tsx (hunk 2 above)
Comment (EvanPurkhiser) [reply src/sentry/static/sentry/app/data/forms/accountNotificationSettings.tsx]: Maybe we should mass update some of these, not sure if there's _that_ much win though over just typing the slightly longer version.

**Pull request merged**